sql/core/src/test/scala/org/apache/spark/sql/sources/v2/SimpleWritableDataSource.scala

```diff
@@ -0,0 +1,254 @@
+    // We can't get the real spark job id here, so we use a timestamp and random UUID to simulate
+    // a unique job id.
+    private val jobId = new SimpleDateFormat("yyyyMMddHHmmss", Locale.US).format(new Date()) +
+      "-" + UUID.randomUUID()
+
+    override def createWriterFactory(): DataWriterFactory[Row] = {
+      new SimpleCSVDataWriterFactory(path, jobId, new SerializableConfiguration(conf))
+    }
+
+    override def commit(messages: Array[WriterCommitMessage]): Unit = {
+      val finalPath = new Path(path)
+      val jobPath = new Path(new Path(finalPath, "_temporary"), jobId)
+      val fs = jobPath.getFileSystem(conf)
+      try {
+        for (file <- fs.listStatus(jobPath).map(_.getPath)) {
+          fs.rename(file, new Path(finalPath, file.getName))
+        }
+      } finally {
+        fs.delete(jobPath, true)
+      }
+    }
+
+    override def abort(messages: Array[WriterCommitMessage]): Unit = {
+      val jobPath = new Path(new Path(path, "_temporary"), jobId)
+      val fs = jobPath.getFileSystem(conf)
+      fs.delete(jobPath, true)
+    }
+  }
+
+  class InternalRowWriter(path: String, conf: Configuration)
+    extends DataSourceV2Writer with SupportsWriteInternalRow {
+
+    private val jobId = new SimpleDateFormat("yyyyMMddHHmmss", Locale.US).format(new Date())
+
+    override def createInternalRowWriterFactory(): DataWriterFactory[InternalRow] = {
+      new InternalRowCSVDataWriterFactory(path, jobId, new SerializableConfiguration(conf))
+    }
+
+    override def commit(messages: Array[WriterCommitMessage]): Unit = {
+      val finalPath = new Path(path)
+      val jobPath = new Path(new Path(finalPath, "_temporary"), jobId)
+      val fs = jobPath.getFileSystem(conf)
+      try {
+        for (file <- fs.listStatus(jobPath).map(_.getPath)) {
+          fs.rename(file, new Path(finalPath, file.getName))
+        }
+      } finally {
+        fs.delete(jobPath, true)
+      }
+    }
+
+    override def abort(messages: Array[WriterCommitMessage]): Unit = {
+      val jobPath = new Path(new Path(path, "_temporary"), jobId)
+      val fs = jobPath.getFileSystem(conf)
+      fs.delete(jobPath, true)
+    }
+  }
+
+  override def createReader(options: DataSourceV2Options): DataSourceV2Reader = {
+    val path = new Path(options.get("path").get())
+    val conf = SparkContext.getActive.get.hadoopConfiguration
+    new Reader(path.toUri.toString, conf)
+  }
+
+  override def createWriter(
+      schema: StructType,
+      mode: SaveMode,
+      options: DataSourceV2Options): Optional[DataSourceV2Writer] = {
+    assert(DataType.equalsStructurally(schema.asNullable, this.schema.asNullable))
+    assert(!SparkContext.getActive.get.conf.getBoolean("spark.speculation", false))
+
+    val path = new Path(options.get("path").get())
+    val internal = options.get("internal").isPresent
+    val conf = SparkContext.getActive.get.hadoopConfiguration
+    val fs = path.getFileSystem(conf)
+    if (fs.exists(path)) {
```

Comment:
If you want max efficiency on object stores, you can skip the exists() check in overwrite & ignore; it's only for ErrorIfExists & Appeend where you need to do the check.  At the very least, overwrite can do the delete() without the existence check.

Reply:
it's just a test and we may not need to worry about performance here.

Reply:
As long as its not the code which ends up being used as the example template, no, nothing to care about.

Reply:
I take this point, this test may eventually become the example template. I'll update it.